ntbs-service/Jobs/DataQualityAlertsJob.cs

```diff
@@ -106,6 +108,15 @@ public async Task Run(IJobCancellationToken token)
             await CreateTreatmentOutcome24MonthAlertsInBulkAsync();
             await CreateTreatmentOutcome36MonthAlertsInBulkAsync();
 
+            // When we check for treatment outcome alerts above, quite a lot of notifications are being brought
+            // into the context. (This is because the eligibility check cannot be entirely translated into SQL so
+            // we are fetching more notifications that we really need from the database and processing them in
+            // memory). Because Notifications have quite a lot of "owns" relationships, this is bringing quite a
+            // large number of objects into the context. This in turn can make save operations slow. We have seen
+            // that each duplicate alert save below was taking more than a second. Here we clear the change tracker
+            // meaning that the context is able to handle saves much more quickly.
+            _context.ChangeTracker.Clear();
+
```

Comment:
This makes perfect sense when you explain it here, but I'm in very impressed that you managed to work out that this was the problem. What made you think about this?

Reply:
Using the debugger, I was able to determine that is was the call to `SaveChanges` which was slow. Based on your earlier investigation which found that the database calls are not slow, that meant that it had to be something internal to EF. In the debugger you can also see how many objects are stored in the local context - and I could see a large number of notifications (and their navigation properties). A large context slowing things down is something that I've run into before using EF, so I guessed that that was the problem, and put this line in to see if it did indeed speed things up (which it did). So a bit of a lucky guess really(!) based off seeing similar kinds of problems before.

Reply:
Thanks for explaining, that's super interesting!